File: pippip-native-ios/NewAccountCreator.m
//
//  NewAccountCreator.m
//  pippip-native-ios
//
//  Created by Steve Brenneis on 10/10/17.
//  Copyright © 2017 seComm. All rights reserved.
//

#import "NewAccountCreator.h"
#import "NewAccountRequest.h"
#import "NewAccountResponse.h"
#import "NewAccountFinish.h"
#import "NewAccountFinal.h"
#import "AlertErrorDelegate.h"
#import "ParameterGenerator.h"
#import "AppDelegate.h"
#import "AccountSession.h"
#import "UserVault.h"
#import "CKPEMCodec.h"

typedef enum STEP { REQUEST, FINISH } ProcessStep;

@interface NewAccountCreator ()
{
    ProcessStep step;
    SessionState *sessionState;
    NSString *passphrase;
}

@property (weak, nonatomic) HomeViewController *viewController;
@property (weak, nonatomic) RESTSession *session;

@end

@implementation NewAccountCreator

@synthesize errorDelegate;
@synthesize postPacket;

- (instancetype)initWithViewController:(HomeViewController *)controller withRESTSession: (RESTSession*)restSession {
    self = [super init];

    _viewController = controller;
    errorDelegate = [[AlertErrorDelegate alloc] initWithViewController:_viewController
                                                             withTitle:@"New Account Creation Error"];
    _session = restSession;

    return self;

}

- (void)accountCreated {

    [self storeVault];
    sessionState.authenticated = YES;
    dispatch_async(dispatch_get_main_queue(), ^{
        AppDelegate *delegate = (AppDelegate*)[[UIApplication sharedApplication] delegate];
        [delegate.accountManager setDefaultConfig];
        [delegate.accountManager storeConfig:sessionState.currentAccount];
        [delegate.accountSession startSession:sessionState];
    });

}

- (void)createAccount:(NSString*)accountName withPassphrase:(NSString *)pass {

    passphrase = pass;
    [_viewController updateStatus:@"Generating user data"];
    ParameterGenerator *generator = [[ParameterGenerator alloc] init];
    [generator generateParameters:accountName];
    sessionState = generator;
    // Start the session.
    [_viewController updateStatus:@"Contacting the message server"];
    [_session startSession:self];

}

- (void) doFinish {

    step = FINISH;
    postPacket = [[NewAccountFinish alloc] initWithState:sessionState];
    [_viewController updateStatus:@"Finishing account creation"];
    [_session queuePost:self];

}

// This will be called after the new account request completes
- (void)postComplete:(NSDictionary*)response {

    if (response != nil) {
        switch (step) {
            case REQUEST:
                if ([self validateResponse:response]) {
                    [self doFinish];
                }
                break;
            case FINISH:
                if ([self validateFinish:response]) {
                    [self accountCreated];
                    [_viewController authenticated:@"Account created. Online"];
                }
                break;
        }

    }

}

- (void)sessionComplete:(NSDictionary*)response {

    if (response != nil) {
        NSString *sessionId = [response objectForKey:@"sessionId"];
        NSString *serverPublicKeyPEM = [response objectForKey:@"serverPublicKey"];
        if (sessionId == nil) {
            [errorDelegate sessionError:@"Invalid server response, missing session ID"];
        }
        else if (serverPublicKeyPEM == nil) {
            [errorDelegate sessionError:@"Invalid server response, missing public key"];
        }
        else {
            sessionState.sessionId = [sessionId intValue];
            CKPEMCodec *pem = [[CKPEMCodec alloc] init];
            sessionState.serverPublicKey = [pem decodePublicKey:serverPublicKeyPEM];
            step = REQUEST;
            postPacket = [[NewAccountRequest alloc] initWithState:sessionState];
            [_viewController updateStatus:@"Requesting account"];
            [_session queuePost:self];
        }
    }

}

- (void)storeVault {
    
    NSArray *paths = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES);
    NSString *docPath = [paths objectAtIndex:0];
    NSString *vaultsPath = [docPath stringByAppendingPathComponent:@"PippipVaults"];
    NSString *vaultPath = [vaultsPath stringByAppendingPathComponent:sessionState.currentAccount];
    
    UserVault *vault = [[UserVault alloc] initWithState:sessionState];
    NSData *vaultData = [vault encode:passphrase];
    
    NSFileManager *fileManager = [NSFileManager defaultManager];
    [fileManager createDirectoryAtPath:vaultsPath
           withIntermediateDirectories:NO
                            attributes:nil
                                 error:nil];
    [vaultData writeToFile:vaultPath atomically:YES];
    
}

- (BOOL) validateFinish:(NSDictionary*)response {

    NewAccountFinal *accountFinal = [[NewAccountFinal alloc] initWithState:sessionState];
    return [accountFinal processResponse:response
                           errorDelegate:errorDelegate];

}

- (BOOL) validateResponse:(NSDictionary*)response {

    NewAccountResponse *accountResponse = [[NewAccountResponse alloc] initWithState:sessionState];
    return [accountResponse processResponse:response
                              errorDelegate:errorDelegate];

}

@end

File: pippip-native-ios/Auth/NewAccountCreator.swift
//
//  NewAccountCreator.swift
//  pippip-native-ios
//
//  Created by Steve Brenneis on 6/1/18.
//  Copyright © 2018 seComm. All rights reserved.
//

import UIKit
import CocoaLumberjack
import Promises

class NewAccountCreator: NSObject {

    var passphrase = ""
    var accountName = ""
    var alertPresenter = AlertPresenter()
    var sessionState = SessionState()
    var biometricsEnabled: Bool!
    var authView: AuthView

    init(authView: AuthView) {

        self.authView = authView
        
        super.init()

    }

    func accountCreated() {

        do {
            try storeVault()
            AccountSession.instance.setDefaultConfig(accountName: accountName)
            let config = Configurator()
            config.useLocalAuth = biometricsEnabled
            config.uuid = NSUUID().uuidString
            if biometricsEnabled {
                if setKeychainPassphrase(uuid: config.uuid , passphrase: passphrase) {
                    authView.accountCreated(success: true, nil)
                    AccountSession.instance.accountCreated(accountName: accountName)
                } else {
                    authView.accountCreated(success: false, "Unable to store passphrase in keychain")
                }
            } else {
                authView.accountCreated(success: true, nil)
            }
        } catch {
            DDLogError("Store user vault error: \(error.localizedDescription)")
            authView.accountCreated(success: false, "Unable to store user vault")
        }

    }

    func createAccount(accountName: String, passphrase: String, biometricsEnabled: Bool) {

        self.passphrase = passphrase
        self.accountName = accountName
        self.biometricsEnabled = biometricsEnabled
        NotificationCenter.default.addObserver(self, selector: #selector(parametersGenerated(_:)),
                                               name: Notifications.ParametersGenerated, object: nil)
        
        DispatchQueue.global().async {
            let parameterGenerator = ParameterGenerator()
            parameterGenerator.generateParameters(accountName)
        }

    }

    func doFinish() {

        let promise: Promise<NewAccountFinal> = SecommAPI.instance.doPost(request: NewAccountFinish())
        promise.then { response in
            self.accountFinishComplete(response)
        }.catch { error in
            self.accountFinishError(error: error.localizedDescription)
        }
        
    }

    func requestNewAccount() {

        let promise: Promise<NewAccountResponse> = SecommAPI.instance.doPost(request: NewAccountRequest())
        promise.then { response in
            self.accountRequestComplete(response)
        }.catch { error in
            self.accountRequestError(error: error.localizedDescription)
        }
        
    }

    func sessionStarted(sessionResponse: SessionResponse) {
        
        if sessionResponse.error != nil {
            alertPresenter.errorAlert(title: "Session Error", message: sessionResponse.error!)
        }
        else {
            sessionState.sessionId = sessionResponse.sessionId!
            let pem = CKPEMCodec()
            sessionState.serverPublicKey = pem.decodePublicKey(sessionResponse.serverPublicKey!)
            self.requestNewAccount()
        }
        
    }
    
    func setKeychainPassphrase(uuid: String, passphrase: String) -> Bool {
        
        let keychain = Keychain(service: Keychain.PIPPIP_TOKEN_SERVICE)
        do {
            try keychain.accessibility(protection: .passcodeSetThisDeviceOnly, createFlag: .userPresence)
                .set(passphrase: passphrase, key: uuid)
            return true
        }
        catch {
            DDLogError("Error adding passphrase to keychain: \(error)")
            self.alertPresenter.errorAlert(title: "Authentication Error", message: "Unable to store passphrase")
            return false
        }
        
    }
    
    func storeVault() throws {

        let paths = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)
        var vaultUrl = paths[0]
        vaultUrl.appendPathComponent("PippipVaults", isDirectory: true)

        let vault = UserVault()
        let vaultData = try vault.encode(passphrase)
        try FileManager.default.createDirectory(at: vaultUrl, withIntermediateDirectories: true, attributes: nil)

        vaultUrl.appendPathComponent(accountName)
        try vaultData.write(to: vaultUrl, options: .atomicWrite)

    }

    // Observer functions

    func accountFinishComplete(_ accountFinal: NewAccountFinal) {

        do {
            try accountFinal.processResponse()
            accountCreated()
        } catch {
            DDLogError("New account finish error: \(error.localizedDescription)")
            alertPresenter.errorAlert(title: "New Account Error", message: error.localizedDescription)
            authView.accountCreated(success: false, error.localizedDescription)
        }
        
    }

    func accountFinishError(error: String) {
        DDLogError("Account finish error: \(error)")
        authView.accountCreated(success: false, error)
    }

    func accountRequestComplete(_ accountResponse: NewAccountResponse) {

        do {
            try accountResponse.processResponse()
            self.doFinish()
        } catch {
            DDLogError("New account request error: \(error.localizedDescription)")
            self.alertPresenter.errorAlert(title: "New Account Error", message: error.localizedDescription)
            self.authView.accountCreated(success: false, error.localizedDescription)
        }
        
    }

    func accountRequestError(error: String) {
        
        DDLogError("Account request error: \(error)")
        authView.accountCreated(success: false, error)
    
    }

    // Notifications

    @objc func parametersGenerated(_ notification: Notification) {

        SecommAPI.instance.startSession(sessionComplete: { (sessionResponse) in
            self.sessionStarted(sessionResponse: sessionResponse)
        })

    }

}
